python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+    activity.logger().debug("Attempt %s", attempts)
+
+
+@activity.run(name="MyActivity")
+async def another_activity(arg1):
+    # Showing heartbeat and cancellation. In Python 3.7 we cannot give a
+    # specific error to Task/Future.cancel (only in 3.9). So activity
+    # cancellations appear like any other asyncio cancellations (which may be a
+    # good thing anyways).
+    try:
+        while True:
+            await asyncio.sleep(1)
+            await activity.heartbeat("some details")
+    except asyncio.CancelledError as exc:
+        raise RuntimeError("someone cancelled me!") from exc
+
+class MyCallableClass:
+    def __call__(self):
+        pass
+
+@activity.run(name="ManualClassName")
+class AnotherCallableClass:
+    def __call__(self):
+        pass
+
+class MyActivities:
+    @activity.run
+    def activity1():
+        pass
+
+    def not_automatically_registered():
+        pass
+
+def register_all(worker):
+    # Registers as "simple_activity"
+    worker.register_activity(simple_activity)
+    # Using customized name
+    worker.register_activity(another_activity)
+    # Registers as "MyCallableClass"
+    worker.register_activity(MyCallableClass())
+    # Registers as "ManualClassName"
+    worker.register_activity(AnotherCallableClass())
+    # Manual name
+    worker.register_activity(another_activity, name="ManualActivityName")
+    # Lambda w/ name
+    worker.register_activity(lambda arg1: arg1, name="MyLambda")
+    # Lambda can of course manually apply decorator
+    worker.register_activity(activity.run(name="MyLambda2")(lambda arg1: arg1))
+    # Methods work just like top-level methods
+    worker.register_activity(MyActivities().not_automatically_registered)
+
+    # Fails because no name provided and it's not a top-level function
+    def nested_activity():
+        pass
+    worker.register_activity(nested_activity)
+    # Fails because it's not callable
+    worker.register_activity(MyActivities())
+
+    # Registers only "activity1"
+    worker.register_activities(MyActivities())
+    # As if calling register_activity on each
+    worker.register_activities([another_activity, MyCallableClass()])
+    # As if calling register_activity with a custom name on each
+    worker.register_activities({'activity_name': simple_activity})
+
+    # Fails, has no @activity.run method
+    worker.register_activities(MyCallableClass())
+```
+
+Notes:
+
+* We use `@activity.run` from `temporalio.activity` instead of just `@activity` from because:
+  * We want `activity` to be a module because we want it usually imported as `from temporalio import activity` so
+    top-level calls like `activity.info()` read well
+  * When we get to workflows, we will do similar with `@workflow.run` vs `@workflow.signal` vs `@workflow.query`, etc
+    and `workflow.info()`
+* Cancellation is task cancellation
+  * Do we need to differentiate worker shutdown from server-side cancellation?
```

Comment:
It would be good to include that information somehow

Reply:
Yes, in addition to a future that they can wait on for cancellation, I will also provide an "is cancelled" bool (or other similar approach to https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.cancelled).

Reply:
> We might be able to use custom tasks so we can throw our own CancelledError subclass.

You have to do more than that. You also have to do your own event loop because you have to be able to intercept https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_future too not just tasks.

Reply:
Added cancelled boolean and a task to wait for cancellation